scripts/licenses/licenses/library.go

```diff
@@ -15,78 +15,133 @@
 package licenses
 
 import (
-	"fmt"
+	"context"
 	"go/build"
+	"path/filepath"
 	"sort"
-	"sync"
+	"strings"
 
 	"github.com/golang/glog"
-)
-
-var (
-	pkgCache sync.Map
+	"golang.org/x/tools/go/packages"
 )
 
 // Library is a collection of packages covered by the same license file.
 type Library struct {
-	Packages    []*build.Package
+	// LicensePath is the path of the file containing the library's license.
 	LicensePath string
+	// Packages contains import paths for Go packages in this library.
+	// It may not be the complete set of all packages in the library.
+	Packages []string
+}
+
+// PackagesError aggregates all Packages[].Errors into a single error.
+type PackagesError struct {
+	pkgs []*packages.Package
+}
+
+func (e PackagesError) Error() string {
+	var errs []string
+	packages.Visit(e.pkgs, nil, func(pkg *packages.Package) {
+		for _, err := range pkg.Errors {
+			errs = append(errs, err.Error())
+		}
+	})
+	return strings.Join(errs, "\n")
 }
 
 // Libraries returns the collection of libraries used by this package, directly or transitively.
 // A library is a collection of one or more packages covered by the same license file.
 // Packages not covered by a license will be returned as individual libraries.
 // Standard library packages will be ignored.
-func Libraries(ctx *build.Context, pkg *build.Package) ([]*Library, error) {
-	pkgs := map[string]*build.Package{pkg.ImportPath: pkg}
-	if err := dependencies(ctx, pkg, pkgs); err != nil {
+func Libraries(ctx context.Context, importPaths ...string) ([]*Library, error) {
+	cfg := &packages.Config{
+		Context: ctx,
+		Mode:    packages.NeedImports | packages.NeedDeps | packages.NeedFiles | packages.NeedName,
+	}
+
+	rootPkgs, err := packages.Load(cfg, importPaths...)
+	if err != nil {
 		return nil, err
 	}
-	pkgsByLicense := make(map[string][]*build.Package)
-	for _, p := range pkgs {
+
+	pkgs := map[string]*packages.Package{}
+	pkgsByLicense := make(map[string][]*packages.Package)
+	errorOccurred := false
+	packages.Visit(rootPkgs, func(p *packages.Package) bool {
+		if len(p.Errors) > 0 {
+			errorOccurred = true
+			return false
+		}
 		if isStdLib(p) {
 			// No license requirements for the Go standard library.
-			continue
+			return false
+		}
+		var pkgDir string
+		switch {
+		case len(p.GoFiles) > 0:
+			pkgDir = filepath.Dir(p.GoFiles[0])
```

Comment:
do you want to concat all these options instead of terminating at the first case?